### DISCUSSION TITLE
Spring boot integrating with last version of parameter-store or secrets (sample)

### DISCUSSION THREAD

**Discussion by hariohmprasath** [awspring/spring-cloud-aws#12] [Q&A] [answered]
Hi,
I am trying to create a spring boot app integrating with the latest version of ```"spring-cloud-aws-parameter-store-config (3.0.0-SNAPSHOT)" module```. Based on the documentation I have defined parameter store entries in "bootstrap.properties" like below:

```properties
aws.paramstore.prefix=/config
aws.paramstore.name=abc
aws.paramstore.enabled=true
aws.paramstore.profileSeparator=_
```

My spring boot main class is just having "@SpringBootApplication" annotation. When I start the application I don't see the parameter values read from AWS. I am using spring boot 2.4.1 like below:
```xml
<parent>
 <groupId>org.springframework.boot</groupId>
 <artifactId>spring-boot-starter-parent</artifactId>
 <version>2.4.1</version>
</parent>
```

If we have any spring boot sample app working with the latest version of "spring-cloud-aws", please share it.

Thank,
Hari

**Comment by eddumelendez** [comment, marked as answer]
You should use `io.awspring.cloud:spring-cloud-starter-aws-secrets-manager-config:2.3.0-RC1` dependency. If you are not using `spring.config.import=aws-parameterstore:` property then use `org.springframework.cloud:spring-cloud-starter-bootstrap` dependency as well. This should be at docs but the site is not available right now

**Reply by hariohmprasath** [reply]
Thanks a lot for your help. I am able to proceed now.

**Comment by petrucristescu** [comment]
Hello,

In pom.xml I have:

 ```
       <dependency>
            <groupId>io.awspring.cloud</groupId>
            <artifactId>spring-cloud-starter-aws-secrets-manager-config</artifactId>
            <version>2.3.0</version>
            <type>pom</type>
        </dependency>
        
    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.4.3</version>
        <relativePath/> <!-- lookup parent from repository -->
    </parent>
```
    
In the application.properties the following:
```
spring.config.import=aws-parameterstore:

aws.paramstore.enabled=true

aws.paramstore.region=region

aws.paramstore.prefix=/config

aws.paramstore.defaultContext=application

aws.paramstore.profileSeparator=_

```

I guess I should use then something like

`@Value("${/config/application/someParam}")`

But what I get is:

`java.lang.IllegalStateException: Unable to load config data from 'aws-parameterstore:'`

Any reason for this?

Thanks in advance

**Reply by eddumelendez** [reply]
@petrucristescu you are using aws-parameter-store prefix but importing secrets-manager starter. Also, yo shoul just use `@Value(someParam)`

**Reply by petrucristescu** [reply]
@eddumelendez Thanks for your fast reply.

Changed to:

```
        <dependency>
            <groupId>io.awspring.cloud</groupId>
            <artifactId>spring-cloud-starter-aws-parameter-store-config</artifactId>
            <version>2.3.0</version>
            <type>pom</type>
        </dependency>
```        
    
```    
spring.config.import=aws-parameterstore:

aws.paramstore.enabled=true

aws.paramstore.region=region

aws.paramstore.prefix=/config

aws.paramstore.defaultContext=application

aws.paramstore.profileSeparator=_
```

```
@Value("/config/application/someValue")
```

In parameter store I have: /config/application/someValue

I still get 
```
java.lang.IllegalStateException: Unable to load config data from 'aws-parameterstore:'
```

I think it's something happening even before reaching any attempt to read the value from parameter store.
Probably I'm missing something else.

Thanks

**Reply by eddumelendez** [reply]
remove type pom from parameter store dependency

**Reply by petrucristescu** [reply]
After removing pom from depenency the "java.lang.IllegalStateException: Unable to load config data from 'aws-parameterstore:'" dissapeared.

But the value from parameter store is not displayed:

What I get is only a echo of the key
```
    @Value("/config/application/somevalue")
    private String paramStoreValue;
```

In the logs I get:
```
The variable from parameter store is:/config/application/somevalue
```

Tried also with:

```
    @Value("somevalue")
    private String paramStoreValue;
```

I get the same result.

**Reply by petrucristescu** [reply]
Nevertheless tried with:
```
@Value("${someValue}")
```

This seems to work. Your help is greatly appreciated.

**Comment by alexdinescu** [comment]
Hello,

in my tests I'd want to disable getting the properties from paramstore via:
`aws.paramstore.enabled=false`
however, if this is added, I'm back to the error:
`Unable to load config data from 'optional:aws-parameterstore:'`
trying to override the `spring.config.import=` to something else in the `test.properties` didn't seem to help

any idea ?

**Reply by maciejwalkowiak** [reply]
You can disable Parameter Store integration through `aws.parameterstore.enabled=false` but then it will fail to resolve `aws-parameterstore:....` config import. Perhaps you should include this import only in the Spring profile that is actually meant to include it?

I assume you want to disable Parameter Store for tests. You can activate "test" profile for all integration test, and include `spring.config.import=aws-parameterstore` only for profile `default` (default is a profile that gets activated when no other profile is active).

**Reply by alexdinescu** [reply]
thanks, that is the solution I've also found and used ( using a 'test' profile ); but I was hoping I could avoid altogether the use of profiles ( previously we didn't need it - with the explicit import bootstrap dependency + bootstrap file ) 

good point about `default` profile, I was using currently something like below in the main `application.properties`, it's cleaner with the default:
```
spring.config.activate.on-profile=!test
spring.config.import=aws-parameterstore:
```

**Reply by maciejwalkowiak** [reply]
There are [ways](https://stackoverflow.com/questions/68702479/disable-config-service-client-activated-with-spring-config-import-in-springboott) to do it without profile but not sure how elegant they are. Anyways, the problem you are having is not specific to Spring Cloud AWS, but rather to Spring Boot's `spring.config.import` - perhaps you can create issue in Spring Boot repository (or maybe there is one already).

**Reply by bigunyak** [reply]
Hi @maciejwalkowiak ,
I'm sorry for resurrecting this one year old thread but I have exactly the same question as @alexdinescu was asking and hope you could help me. 🙏 
I understand the suggestion you gave with a specific `test` profile where `spring.config.import` is not used and thus should prevent ParameterStore integration from loading. But this solution doesn't seem to work with Spring Cloud AWS 2.4.4 that I'm using.
My goal is to use `spring.config.import` feature without `bootstrap.yml` but `aws.paramstore.enabled: false` seems to be only working in the `bootstrap.yml`, where I could have something like this:
```yaml
---
spring.config.activate.on-profile: test
aws.paramstore.enabled: false
```
My guess is because Spring Cloud AWS 2.4.4 is still relying on:
```
org.springframework.cloud.bootstrap.BootstrapConfiguration=\
io.awspring.cloud.autoconfigure.paramstore.AwsParamStoreBootstrapConfiguration
```
Which leads to ParameterStore integration being loaded and me getting this error:
```
io.awspring.cloud.paramstore.AwsParamStorePropertySources$AwsParameterPropertySourceNotFoundException:
 com.amazonaws.services.simplesystemsmanagement.model.AWSSimpleSystemsManagementException:
 The security token included in the request is expired (Service: AWSSimpleSystemsManagement; Status Code: 400; Error Code:
 ExpiredTokenException; Request ID: 3949e500-bfd7-4248-9860-6f080f3fd5ee; Proxy: null)
```
The question is this, in Spring Cloud AWS 2.4.4, can I fully rely on `spring.config.import=aws-parameterstore:` without using `bootstrap.yml` and disable ParameterStore integration from loading in unit tests?

I'd very much appreciate your help.

**Reply by alexdinescu** [reply]
hi @bigunyak ,

one other option I've ended up using is the one explained [here](https://stackoverflow.com/a/69739888/981788):

in the main `application.properties` use something like:
```
config.type=aws-parameterstore:
spring.config.import=${config.type}
```
then, in your tests properties, override the `config.type` as well as disabling param store:
```
config.type=

aws.paramstore.enabled=false
```

hope this helps

**Reply by dominik-kovacs** [reply]
@alexdinescu this does not work for me :(

**Reply by bigunyak** [reply]
Thanks for your suggestion, @alexdinescu.
The issue with the spring-cloud-aws v2.4.x is that it seems to be a weird mix of two implementations, bootstrap-based and `spring.config.import`-based, and the way they intersect is not at all obvious. The documentation also doesn't providing clear guidelines on how to migrate from bootstrap-based implementation to `spring.config.import`.
From digging into the code, it doesn't look that it's possible to get rid of `bootstrap.yml` config with 2.4.x version. So, I ended up with the following approach:
1. For Parameter Store, we'll keep using `bootstrap.yml` based setup, until we upgrade to SpringBoot 3.0 and spring-cloud-aws 3.0 where bootstrap support was removed.
2. For Secrets Manager, we found a way to utilize `spring.config.import` in v2.4.4, but it only works with `aws.secretsmanager.failFast: false` set in `bootstrap.yml`, which essentially ignores all not found secrets which `bootstrap`-based implementation is trying to find.
